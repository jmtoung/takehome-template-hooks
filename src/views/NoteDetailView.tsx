import React from 'react';
import { Container, Jumbotron } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router-dom';
import { NoteForm } from '../components/NoteForm';
import { API } from '../api';
import { NotFound } from '../components/NotFound';
import { useNote } from '../hooks/useNote';

type TParams = { id: string };

export const NoteDetailView = ({
  match,
  history,
}: RouteComponentProps<TParams>) => {
  const { id } = match.params;
  const { title, body, setTitle, setBody, notFound } = useNote(id);

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    await API.updateNote(id, title, body);
    history.goBack();
  };

  if (notFound) {
    return <NotFound />;
  }

  if (title === '' && body === '') {
    return null;
  }

  return (
    <Container>
      <Jumbotron>
        <h1>Note Editor</h1>
        <NoteForm
          text="Update Note"
          title={title}
          body={body}
          handleSubmit={(e) => handleSubmit(e)}
          handleBodyChange={(e) => setBody(e.target.value)}
          handleTitleChange={(e) => setTitle(e.target.value)}
        />
      </Jumbotron>
    </Container>
  );
};
