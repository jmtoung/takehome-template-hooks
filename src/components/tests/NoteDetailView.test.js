import React from 'react';
import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import NoteDetailView from '../../views/NoteDetailView';

describe('NoteDetailView', () => {
  it('renders NotFound component if no id.', async (done) => {
    jest.spyOn(React, 'useEffect').mockImplementation(f => f());
    const wrapper = mount(<NoteDetailView match={{ params: { id: null} }} />);


    expect(React.useEffect).toHaveBeenCalledTimes(1);
    expect(wrapper.props()).toEqual(true);


    done();
  });
  // it('renders NotFound component if no id.', async () => {
  //   jest.spyOn(React, 'useEffect').mockImplementation(f => f());
  //   const wrapper = mount(<NoteDetailView match={{ params: { id: null} }} />);
  //   expect(React.useEffect).toHaveBeenCalledTimes(1);
  //   expect(toJSON(wrapper)).toMatchSnapshot();
  // });

  // it('renders properly after fetching note.', async (done) => {
  //   const mockSuccessResponse = { title: 'This is a title', body: 'This is the body.' };
  //   const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  //   const mockFetchPromise = Promise.resolve({
  //     json: () => mockJsonPromise,
  //   });
  //   jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

  //   const wrapper = shallow(<NoteDetailView match={{ params: { id: 1 } }} />);
  //   expect(global.fetch).toHaveBeenCalledTimes(1);
  //   expect(global.fetch).toHaveBeenCalledWith('https://5ec581652a4ba000163d2084.mockapi.io/api/v1/notes/1');

  //   process.nextTick(() => {
  //     expect(wrapper.state()).toEqual({
  //       title: 'This is a title',
  //       body: 'This is the body.',
  //       notFound: false,
  //     });
  //     expect(toJSON(wrapper)).toMatchSnapshot();
  //   });

  //   global.fetch.mockClear();
  //   done();
  // });


});