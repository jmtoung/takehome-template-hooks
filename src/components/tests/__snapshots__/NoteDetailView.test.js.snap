// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteDetailView renders NotFound component if no id. 1`] = `""`;
